cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/JarCacheTracker.java

```diff
@@ -0,0 +1,179 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Optimization for CDAP-7021.
+ *
+ * We cache appmaster and container jars created by twill the first time a program
+ * of that type is run. However, if 100 programs are started at the same time, all 100 will still create those
+ * jars since the cached jars won't be available at run start. So as an optimization, if a program is
+ * launching and its cached jars are not available, and its not the first program of its type to be launched,
+ * and its launching within 90 seconds of the first run, wait for the cache jars to become available.
+ * Without this, we might actually miss schedule runs due to jar building taking too much time and resources.
+ */
+public class JarCacheTracker {
+  public static final JarCacheTracker INSTANCE = new JarCacheTracker();
+  private static final Logger LOG = LoggerFactory.getLogger(JarCacheTracker.class);
+  private final Map<ProgramType, CacheInfo> programCaches;
+
+  private JarCacheTracker() {
+    programCaches = new ConcurrentHashMap<>();
+    for (ProgramType programType : ProgramType.values()) {
+      programCaches.put(programType, new CacheInfo());
+    }
+  }
+
+  public void registerLaunch(File cacheDir, ProgramType programType) {
+    CacheInfo cacheInfo = programCaches.get(programType);
+
+    long launchTime = System.currentTimeMillis();
+    if (cacheInfo.firstLaunchTime.compareAndSet(-1L, launchTime)) {
+      // if this is the first program of this type to launch, no need to wait
+      LOG.debug("First program of type {} launched at time {}", programType, launchTime);
+      return;
+    }
+
+    // if we're more than 90 seconds past the first launch time, no need to waits
+    long threshold = cacheInfo.firstLaunchTime.get() + TimeUnit.MILLISECONDS.convert(90, TimeUnit.SECONDS);
+    if (System.currentTimeMillis() > threshold) {
+      return;
+    }
+
+    // otherwise, wait until at most 90 seconds past the first launch time for the cache files to exist
+    cacheInfo.waitForCacheFiles(cacheDir);
+  }
+
+  /**
+   * Used to hold the first time a program of the type was launched and wait for the cache files to exist.
+   */
+  private class CacheInfo {
+    private final AtomicLong firstLaunchTime = new AtomicLong(-1L);
+
+    // synchronized here instead of at registerLaunch so that we only per program type.
+    private synchronized void waitForCacheFiles(File cacheDir) {
+      // check this in case we called this method before the files were available, were blocked on another thread,
+      // and are now here
+      long threshold = firstLaunchTime.get() + TimeUnit.MILLISECONDS.convert(90, TimeUnit.SECONDS);
+      if (System.currentTimeMillis() > threshold || cacheFilesExist(cacheDir)) {
+        return;
+      }
+
+      // otherwise, wait for at most threshold seconds for the cache files to exist
+      WatchService watcher;
+      WatchKey key;
+      try {
+        watcher = FileSystems.getDefault().newWatchService();
```

Comment:
`WatchService` is a closeable, need to be closed after used. Otherwise it will leak fd. Need to wrap it with `try () { }`.